Documentation/botr/readytorun-format.md

```diff
@@ -326,6 +382,49 @@ This section marks region that contains thunks for ``READYTORUN_HELPER_DelayLoad
+The version-resilient hashing algorithm used for hashing the type names is implemented in [vm/versionresilienthashcode.cpp](https://github.com/dotnet/coreclr/blob/ec2a74e7649f1c0ecff32ce86724bf3ca80bfd46/src/vm/versionresilienthashcode.cpp#L75).
+
+## READYTORUN_SECTION_INSTANCE_METHOD_ENTRYPOINTS
+
+This section contains a native hashtable of all generic method instantiations compiled into the R2R executable. The key is the method instance signature; the appropriate version-resilient hash code calculation is implemented in [vm/versionresilienthashcode.cpp](https://github.com/dotnet/coreclr/blob/ec2a74e7649f1c0ecff32ce86724bf3ca80bfd46/src/vm/versionresilienthashcode.cpp#L128); the value, represented by the ``EntryPointWithBlobVertex`` class, stores the method index in the runtime function table, the fixups blob and a blob encoding the method signature.
+
+## READYTORUN_SECTION_INLINING_INFO
+
+**TODO**: document inlining info encoding
+
+## READYTORUN_SECTION_PROFILEDATA_INFO
+
+**TODO**: document profile data encoding
+
+## READYTORUN_SECTION_MANIFEST_METADATA
+
+Manifest metadata is an [ECMA-335] metadata blob containing extra reference assemblies introduced by inlining on top of assembly references stored in the input MSIL. This can only happen for non-versionable methods or for methods within the same *version bubble*.
+
+As of R2R version 3.0, The metadata is only searched for the AssemblyRef table. This is used to translate module override indices in signatures to the actual reference modules (using either the ``READYTORUN_FIXUP_ModuleOverride`` bit flag on the signature fixup byte or the ``ELEMENT_TYPE_MODULE_ZAPSIG`` COR element type).
+
+The module override index translation algorithm is as follows (**ILAR** = *the number of `AssemblyRef` rows in the input MSIL*):
+
+| Module override index (*i*) | Reference assembly                                                                               |
+|:----------------------------|:-------------------------------------------------------------------------------------------------|
+| *i* = 0                     | Global context - assembly containing the signature                                               |
+| 1 <= *i* <= **ILAR**        | *i* is the index into the MSIL `AssemblyRef` table                                               |
+| *i* > **ILAR**              | *i* - **ILAR** - 1 is the zero-based index into the `AssemblyRef` table in the manifest metadata |
+
+**Note:** This means that the entry corresponding to *i* = **ILAR** + 1 is actually undefined as it corresponds to the `NULL` entry (ROWID #0) in the manifest metadata AssemblyRef table. The first meaningful index into the manifest metadata, *i* = **ILAR** + 2, corresponding to ROWID #1, is historically filled in by Crossgen with the input assembly info but this shouldn't be depended upon, in fact the input assembly is useless in the manifest metadata as the module override to it can be encoded by using the special index 0.
```

Comment:
The crossgen implementation quirks should not be part of the file format format spec.

Reply:
Hmm, I would rather say that this is a runtime contract regarding how the external module addressing is designed. There's no way to encode non-null row #0 in the manifest metadata as the ECMA335 metadata format simply doesn't support that. So the runtime will always end up with a null row when querying ILAR + 1 no matter whether the executable was produced by Crossgen, CPAOT or any other compiler conforming to the current format design. Could you please suggest what you would perceive as an appropriate way to treat this in the format doc?

Reply:
We just took what happened to be there for fragile NGen. The implementation does not quite make sense as many things in fragile NGen.

We are not actually shipping this as officially supported because of the implementation is incomplete. Maybe the manifest metadata should be marked as work-in-process on future version here. We may want to change and clean this up so that it make sense.

Reply:
I have added a disclaimer echoing your feedback to the doc in 10th commit.